### PR TITLE
reduce risk of corrupt radio.yml #1412

### LINKED ISSUE #1412: reduce risk of corrupt radio.yml
The radio.yml is written relatively often. When the write does not finish the configuration may be broken, depending on the amount of data already written, in ways hard to detect. 
Instead of truncating the file at every write I propose this procedure:
1. open radio_new.yml (or another meaningful name
2. complete writing to the new file, including a checksum
3. delete old file, rename new file, renaming is (normally an atomic operation on file systems)

for reading:
1. try to open regular config file (radio.yml)
2. if it exists, check the checksum
3. if it does not exist or checksum missmatch, try to open the "radio_new.yml" file
4. if it exists, check the checksum
5. if the checksum of either file is good, just run normally
6. if no checksum is OK warn the user about a broken configuration and revert to defaults

At the beginning of the Yaml file an entry "manuallyEdited" should be added, to skip the checksum check (see discussion below, thx @rotorman).

In any case a broken radio.yml should be preserved by copying to a new name, to be analysed later.

I think some of the problems we currently see sometimes are coming from partially written config files.

### PR DESCRIPTION
**Pull request by stephenaa**
Fixes #1412 

Summary of changes:
Adds CRC16 checksum to radio.yml when saved.
Adds save to temporary file, folowed by rename to radio.yml
Adds check on load, with fall-back to temporary file

### CODE REVIEW

File: radio/src/storage/sdcard_yaml.cpp [hunk 1]
```diff
@@ -47,23 +49,62 @@ const char * readYamlFile(const char* fullpath, const YamlParserCalls* calls, vo
...
+
+        // Skip LF of line-ending CR-LF pair
+        endPos++;
+        if(*endPos == '\n') {
+          endPos++;
+        }
+        file_checksum = atoi(startPos);
+        skip = endPos - buffer;
+      }
+
```
Comment (gagarinlg) [inline radio/src/storage/sdcard_yaml.cpp:88]: This should be done only once and not on every read. When longer texts are allowed combined with some bad luck, "checksum: " can be part of another config and by chance be at the beginning of a buffer.
I would prefer a read before you enter the while loop or at least a bool to only check for the checksum at the first run.

File: radio/src/storage/sdcard_yaml.cpp [hunk 2]
```diff
@@ -92,15 +133,43 @@ void storageCreateModelsList()
 // SDCARD storage interface
 //
 
+static const char * attemptLoad(const char *filename, uint16_t* checksum_status)
```
Comment (gagarinlg) [inline radio/src/storage/sdcard_yaml.cpp:136]: Same here as before, I would prefer references for checksum_status.

**Review by gagarinlg — CHANGES_REQUESTED** [review]
Good work, except for some minor things. Looking forward to test it.

File: radio/src/storage/sdcard_yaml.cpp (hunk 1 above)
Comment (stephenaa) [reply radio/src/storage/sdcard_yaml.cpp:88]: Yeah, absolutely right - I will update the PR

**Comment by stephenaa** [comment]
I guess the test will also fail on the coming updated PR : https://github.com/EdgeTX/edgetx/issues/1412

Should I modify the radio.config used for the tests? Or is that provided by the test action setup? 
In any case, if neither checksum nor "manuallyEdited" is present in the file, loading will result in an alert - that causes the test to fail. That needs to be addressed somehow :-)

File: radio/src/storage/sdcard_yaml.cpp (hunk 1 above)
Comment (stephenaa) [reply radio/src/storage/sdcard_yaml.cpp:88]: I added a flag to have the check only run once. I preferred that over having two read blocks of the code - which seemed to fragment the read - and I think it also produces a larger binary - however I am willing to be proven wrong :-)

File: radio/src/storage/sdcard_yaml.cpp (hunk 2 above)
Comment (stephenaa) [reply radio/src/storage/sdcard_yaml.cpp:136]: Yeah, but I also want to have an easy way of controlling wether to not readYamlFile() should do the checksum calculation. Currently the pointer to write the result is controlling this.

I will update the PR - hopefully my intention is a bit clearer now. If you have an idea to do this simpler please comment :-)

File: radio/src/storage/sdcard_yaml.cpp [hunk 3]
```diff
@@ -47,23 +49,69 @@ const char * readYamlFile(const char* fullpath, const YamlParserCalls* calls, vo
...
+
+    if (checksum_result != NULL) {
+      if (calculated_checksum == file_checksum) {
+        *checksum_result = CHECKSUM_SUCCESS;
+      } else {
+        *checksum_result = CHECKSUM_FAILED;
+      }
+    } else {
+      *checksum_result = CHECKSUM_NONE;
+    }
```
Comment (gagarinlg) [inline radio/src/storage/sdcard_yaml.cpp:113]: I think line 112 is not a good idea. Dereferencing a NULL pointer.

File: radio/src/storage/sdcard_yaml.cpp (hunk 2 above)
Comment (gagarinlg) [reply radio/src/storage/sdcard_yaml.cpp:143]: Yes you are right, as along as the same function is used for files with and without checksum, this is probably the best way to do it.

**Review by gagarinlg — CHANGES_REQUESTED** [review]
The file reading is better now.
Please change the CHECKSUM_ defines to an enum or enum class.

File: radio/src/storage/sdcard_yaml.cpp (hunk 3 above)
Comment (stephenaa) [reply radio/src/storage/sdcard_yaml.cpp]: Good catch

Comment (stephenaa) [reply radio/src/storage/sdcard_yaml.cpp]: Update on its way with null-pointer dereference fixed and enum class for return values.

**Comment by stephenaa** [comment]
Changed to use an enum class for the checksum return values.

**Review by gagarinlg — COMMENTED** [review]
Much better, thx

**Review by gagarinlg — APPROVED** [review]

**Comment by Nerenye** [comment]
Just out of curiosity. 

How does it prevent file corruption? AS far as i unterstand it should error out only on missing Checksum, right?
If checksum is present but changed, how do we know that the user didn´t change anything? And how do we know that the content of the yml and therefore the new checksum is correct before we overwrite the old probably correct file?.

i´m probably missing something, really just curious

**Comment by stephenaa** [comment]
It tries to detect if radio.yml was not fully written to flash, and then reverts to the alternative file.
When radio.yml is written it is first saved under the alternative name, then renamed.
So the checksum allows to detect corruption, and gives the SW one shot in reverting to a valid file.

For the user editing manually, there is a YAML variable (manuellyEdited) that makes the SW skip checksum check at load.

**Comment by Nerenye** [comment]
yeah that´s the part i get. but the checksum would also change if you change something in radio configuration e.g not manually editing the file.
Wouldn´t the checksum check fail at that point? or how do we check if the file is corrupted this time. 
I mean if you change somethin in the radio config and a write is triggered, And unlucky as we are the write is corrupted. How does the checksum prevent us from overwriting the old file. SInce as far as i understand it can´t know that the new checksum is not correct ?

**Comment by gagarinlg** [comment]
It writes the new configuration to a different name and only after completely writing, the old file is replaced. So that, at all times, a good woring config file should be there.

**Comment by Nerenye** [comment]
yeah but how do we know it´s a good working file. 

If we change the radio config the checksum changes aswell
if it writes the temp file and the write is corrupted it won´t realise the file is corrupted because its valid that the checksum missmatches because we changed some config entrys.
And will probably overwrite the non corrupted file.

**Comment by gagarinlg** [comment]
The checksum is regenerated with every config change and written as part of the file. It is not static. A config file must either contain the manually edited flag or a correct checksum, that matches the file contents.

**Comment by Nerenye** [comment]
so the checksum is generated before we write and we would know if the write corrupted if we check the file content checksum before we rename.

ok get it. thanks for the clarification 👍

**Comment by stephenaa** [comment]
We don't do a finishing read-back before rename at this point, but that could easily be added. However, do we want to make save take double the time? On all radios?

**Comment by gagarinlg** [comment]
I do not think, that we need this.
If someone wnats to have additional safety here, I would prefer to reneame the old file and then the new file after each write, so that we have the old one as a backup. But the cases where the config is corrupted should be really rare, except when debugging and constantly resetting the radio. In my opinion we are good.

**Comment by gagarinlg** [comment]
I will test this PR at the weekend

**Comment by gagarinlg** [comment]
`-rw-r--r--  1 xxx xxx 2900 Jan 28  2023 radio_error.yml
-rw-r--r--  1 xxx xxx 2933 Jan 29  2022 radio_new.yml
-rw-r--r--  1 xxx xxx 2329 Jan 29  2022 radio.yml
`
I get this after calibration. The old config is still there and the new one is written, but not renamed. Do I have to do something speacial?

**Comment by gagarinlg** [comment]
Did you test this on a reald radio? Maybe renaming when there already is a file does not work.

File: radio/src/storage/sdcard_yaml.cpp [hunk 4]
```diff
@@ -92,15 +138,44 @@ void storageCreateModelsList()
...
+      TRACE("radio settings: checksum check failed");
+      if (g_eeGeneral.manuallyEdited) {
+        TRACE("File has been manually edited - ignoring checksum mismatch");
+        g_eeGeneral.manuallyEdited = 0;
+        storageDirty(EE_GENERAL);   // Trigger a save on sucessfull recovery
+      } else {
+        TRACE("File is corrupted, attempting alternative file");
+        result = f_rename(RADIO_SETTINGS_YAML_PATH, RADIO_SETTINGS_ERRORFILE_YAML_PATH); // Save corrupted file for later analysis
+        p = attemptLoad(RADIO_SETTINGS_TMPFILE_YAML_PATH, &checksum_status);
+        if (p == NULL) {
```
Comment (gagarinlg) [inline radio/src/storage/sdcard_yaml.cpp:167]: somehow I always get an unrecoverable error, when I only have a radio_new.yaml, but no radio.yml

**Review by gagarinlg — CHANGES_REQUESTED** [review]
Some more things I found, while testing with a TX16S

**Comment by stephenaa** [comment]
Nice find - thanks.

I am currently down with Corona - hopefully I am back in cation within a few days :-/

File: radio/src/storage/sdcard_yaml.cpp [hunk 5]
```diff
@@ -92,15 +138,44 @@ void storageCreateModelsList()
...
+    if(checksum_status != ChecksumResult::Success) {
+      FRESULT result = FR_OK;
+      TRACE("radio settings: checksum check failed");
+      if (g_eeGeneral.manuallyEdited) {
+        TRACE("File has been manually edited - ignoring checksum mismatch");
+        g_eeGeneral.manuallyEdited = 0;
+        storageDirty(EE_GENERAL);   // Trigger a save on sucessfull recovery
+      } else {
+        TRACE("File is corrupted, attempting alternative file");
+        result = f_rename(RADIO_SETTINGS_YAML_PATH, RADIO_SETTINGS_ERRORFILE_YAML_PATH); // Save corrupted file for later analysis
```
[the remark this answers is not in the document]
Comment (stephenaa) [reply radio/src/storage/sdcard_yaml.cpp:165]: Should we reconsider the strategy, and do a regular backup-file scheme as was discussed earlier in the comments?

If we need two filesystem operations, the premise of atomicity in the renaming is lost.

[the remark this answers is not in the document]
Comment (gagarinlg) [reply radio/src/storage/sdcard_yaml.cpp:165]: As long as the final rename is atomic, it should not be a problem, When there is no radio.yml, we can load radio_new.yaml and that's it.

**Comment by gagarinlg** [comment]
> I am currently down with Corona - hopefully I am back in cation within a few days :-/

I hope you wil recover quick and without any permanent problems. There is no time preasure, so dion't worry about this.

**Comment by pfeerick** [comment]
> I am currently down with Corona - hopefully I am back in cation within a few days :-/

Take care mate!

File: radio/src/storage/sdcard_yaml.cpp (hunk 4 above)
Comment (stephenaa) [reply radio/src/storage/sdcard_yaml.cpp:167]: I assume this is due to the missing check in line 154 for the reading to fail.
I am working on a fix right now :-)

Comment (stephenaa) [reply radio/src/storage/sdcard_yaml.cpp:167]: Looking at the code I realized a missing check for the temp file's checksum to be correct. I am adding that to my update and push it. I will try to flash a radio tomorrow.

File: radio/src/storage/sdcard_yaml.cpp [hunk 6]
```diff
@@ -92,15 +138,44 @@ void storageCreateModelsList()
...
     TRACE("YAML radio settings reader");
 
-    YamlTreeWalker tree;
-    tree.reset(get_radiodata_nodes(), (uint8_t*)&g_eeGeneral);
-
-    return readYamlFile(RADIO_SETTINGS_YAML_PATH, YamlTreeWalker::get_parser_calls(), &tree);
+    ChecksumResult checksum_status;
+    const char* p = attemptLoad(RADIO_SETTINGS_YAML_PATH, &checksum_status);
+
+    if(checksum_status != ChecksumResult::Success) {
```
[the remark this answers is not in the document]
Comment (stephenaa) [reply radio/src/storage/sdcard_yaml.cpp]: Updated in latest revision

File: radio/src/storage/sdcard_yaml.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (stephenaa) [reply radio/src/storage/sdcard_yaml.cpp:168]: Done

**Comment by gagarinlg** [comment]
I will take a look and test it within the next days. Thank you very much for your work.

**Review by gagarinlg — APPROVED** [review]

**Comment by gagarinlg** [comment]
This works much better now. One thing I found is, that an old config file will trigger an unrecoverable error and reset the configuration. It probably make sense to check, if there is a checksum at the beginning, and if not, but the file has a reasonable size, load it anyways. Or we need a new storage file version to detect this.
@pfeerick, @raphaelcoeffic  what do you think?

also the companion needs to be checksum aware

**Comment by stephenaa** [comment]
Should I add checksum support to companion?

@gagarinlg

**Comment by gagarinlg** [comment]
It would be great if you could add this to companion. In the first step I forgot, because I don't really use companion.

I think the storage version is the only way to detect this.

**Comment by stephenaa** [comment]
> It would be great if you could add this to companion. In the first step I forgot, because I don't really use companion.
> 
> I think the storage version is the only way to detect if the file needs to be updated

I'am on it :-) I was wondering, what should companion behaviour be if a radio.yml is opened, but checksum is wrong? Can we rely on companion to create a valid settings file from scratch? This overlaps a bit with #1521 IMHO.

For the other issue I have been thinking a bit more. Absence of a value "checksum" is only a problem if it is not on the first line, that is - if there is no 'checksum' value, and there is something sensible in the file (and the length is realistic as you suggested earlier) the file is probably (nearly 100%) an old file from before the checksum check was added.

**Comment by stephenaa** [comment]
I found a way around the newline issue. In reality it is only needed when ignoring the checksum field added as the first line of the file. By rearranging the logic I managed to minimize the amount of code exposed to the variability.

Also changed the radio firmware to accept files with no checksum field if their size is >25 bytes. I can't really see a better way right now to qualify an "old file", that is a file from before the checksum field was added. Maybe it should be kept in mind for a future file version version bump?

**Comment by gagarinlg** [comment]
> Also changed the radio firmware to accept files with no checksum field if their size is >25 bytes. I can't really see a better way right now to qualify an "old file", that is a file from before the checksum field was added. Maybe it should be kept in mind for a future file version version bump?

I think the limit could be a lot higher than 25 bytes. Other than that, the code looks good, I will test it tomorrow.

**Comment by stephenaa** [comment]
Keep in mind that the same code reads model YAMl, so the minimum size could have to be the minimum model file if we are to extend checksum to models. 


Sendt fra min iPhone

**Comment by gagarinlg** [comment]
I think the minimum model ist larger, but that as a question for @raphaelcoeffic .
Reagrading checksums on models, I am not sure. Models are not written as often. On the other hand a warning, when a model might be corrupted doesnÄt sound too bad. But this would be a  task for another PR.

**Comment by stephenaa** [comment]
> I think the minimum model ist larger, but that as a question for @raphaelcoeffic .
> Reagrading checksums on models, I am not sure. Models are not written as often. On the other hand a warning, when a model might be corrupted doesnÄt sound too bad. But this would be a task for another PR.

Let's leave the model part for now. Since the function is not called with a pointer for the CRC in case of model files it is a non-issue for now.

**Comment by gagarinlg** [comment]
I found an issue. When there is no radio.yml, but a radio.bin and a radio_new.yml a new bin->yaml conversion is started

**Comment by stephenaa** [comment]
So that conversion is triggered before reading of radio.yml of course. The check should probarbly include “radio_new.yml” as Well?

Sendt fra min iPhone

**Comment by gagarinlg** [comment]
> So that conversion is triggered before reading of radio.yml of course. The check should probarbly include “radio_new.yml” as Well?

Yes, that should fix that problem.

**Comment by stephenaa** [comment]
> > So that conversion is triggered before reading of radio.yml of course. The check should probarbly include “radio_new.yml” as Well?
> 
> Yes, that should fix that problem.

Is that in conversions_220_221.cpp, line 185? Would that be the conversion from bin (220) to YAML (221) ? AS far as my grep-foo goes this is the only candidate to trigger save of a new radio.yml file.

**Comment by stephenaa** [comment]
> RADIO_SETTINGS_YAML_PATH

I guess if there is no radio.yml or radio_new.yml BUT there is a radio_error.yml a new conversion from binary is OK?

**Comment by pfeerick** [comment]
This still has one or two niggling issues, and then I thinks it's ready. 

- When using the below `radio.yml` on my standard test radio, the new unrecoverable data path was followed, rather than assuming it was ok since there was no checksum present, and thus adding it. i.e. the "Special case to handle "old" files with no checksum field" didn't seem to be followed. It seems, on simu at least, that `bytes_read` is zero outside of the `f_read` loop - I'll comment against the code for line reference...
- Companion doesn't know about the manuallyEdited node, so doesn't add it itself, thus the radio will do that after every sync with companion. 

```yml
board: tx16s
version: 221
calib:
  Rud:
    mid: 1021
    spanNeg: 668
    spanPos: 742
  Ele:
    mid: 1045
    spanNeg: 543
    spanPos: 419
  Thr:
    mid: 1043
    spanNeg: 465
    spanPos: 376
  Ail:
    mid: 1038
    spanNeg: 846
    spanPos: 599
  S1:
    mid: 1040
    spanNeg: 1023
    spanPos: 991
  6POS:
    mid: 1029
    spanNeg: 5901
    spanPos: 10784
  S2:
    mid: 1044
    spanNeg: 1027
    spanPos: 987
  EXT1:
    mid: 0
    spanNeg: 0
    spanPos: 0
  EXT2:
    mid: 0
    spanNeg: 0
    spanPos: 0
  EXT3:
    mid: 0
    spanNeg: 0
    spanPos: 0
  EXT4:
    mid: 0
    spanNeg: 0
    spanPos: 0
  LS:
    mid: 1084
    spanNeg: 293
    spanPos: 234
  RS:
    mid: 999
    spanNeg: 243
    spanPos: 237
currModel: 0
currModelFilename: model1.yml
contrast: 25
vBatWarn: 64
txVoltageCalibration: -11
vBatMin: 60
vBatMax: 84
backlightMode: backlight_mode_all
trainer:
  calib:
    0:
      val: -1
    1:
      val: -2
    2:
      val: -12
    3:
      val: 7
  mix:
    0:
      srcChn: 1
      mode: REPL
      studWeight: 100
    1:
      srcChn: 0
      mode: REPL
      studWeight: 100
    2:
      srcChn: 3
      mode: REPL
      studWeight: 100
    3:
      srcChn: 2
      mode: REPL
      studWeight: 100
view: 0
fai: 0
disableMemoryWarning: 0
beepMode: mode_alarms
disableAlarmWarning: 0
disableRssiPoweroffAlarm: 0
USBMode: 0
jackMode: 0
hapticMode: mode_nokeys
stickMode: 1
timezone: 10
adjustRTC: 1
inactivityTimer: 10
telemetryBaudrate: 0
internalModule: TYPE_MULTIMODULE
splashMode: 0
lightAutoOff: 2
templateSetup: 17
hapticLength: 2
speakerPitch: 0
hapticStrength: 0
beepLength: 2
gpsFormat: 1
speakerVolume: 11
backlightBright: 17
blOffBright: 1
switchesDelay: 0
globalTimer: 192967
bluetoothName: ""
bluetoothBaudrate: 0
bluetoothMode: OFF
countryCode: 0
noJitterFilter: 0
disableRtcWarning: 0
keysBacklight: 0
imperial: 0
ttsLanguage: en
beepVolume: 3
wavVolume: 2
varioVolume: 2
varioPitch: 0
varioRange: 0
varioRepeat: 0
backgroundVolume: 3
auxSerialMode: MODE_NONE
aux2SerialMode: MODE_NONE
antennaMode: MODE_PER_MODEL
backlightColor: 0
pwrOnSpeed: 0
pwrOffSpeed: 0
customFn:
  0:
    swtch: ON
    func: BACKLIGHT
    def: S2,1
  1:
    swtch: ON
    func: VOLUME
    def: S1,1
  2:
    swtch: SF2
    func: LOGS
    def: 5
sticksConfig:
  0:
    name: R
  1:
    name: E
  2:
    name: T
  3:
    name: A
switchConfig:
  SA:
    type: 3pos
    name: ""
  SB:
    type: 3pos
    name: ""
  SC:
    type: 3pos
    name: ""
  SD:
    type: 3pos
    name: ""
  SE:
    type: 3pos
    name: ""
  SF:
    type: 2pos
    name: ""
  SG:
    type: 3pos
    name: ""
  SH:
    type: toggle
    name: ""
potsConfig:
  S1:
    type: with_detent
    name: ""
  6POS:
    type: multipos_switch
    name: ""
  S2:
    type: with_detent
    name: ""
slidersConfig:
  LS:
    type: with_detent
    name: ""
  RS:
    type: with_detent
    name: ""
ownerRegistrationID: ""
gyroMax: 0
gyroOffset: 0
uartSampleMode: 0
```

**Comment by pfeerick** [comment]
@elecpower Did you want to add anything about the Companion side of this? It seems to be working ok, just the absence of `manuallyEdited: 0` that I mentioned above.

**Comment by elecpower** [comment]
@pfeerick I'll review and post my comments

**Comment by elecpower** [comment]
**Companion**
Tested by loading an old X9D+ etx with a TX16S profile and simulated.
No checksum message as did not exist - pass
Save converted file and checksum added - pass
Hacked the checksum value, opened the file and checksum error message - pass 
_Comment_
To me the message is not really helpful as it does not indicate which file and the expected value doesn't really matter. If, in future, a checksum is added to each model file then the file name will be essential. How about something like 'Checksum error: possible corruption of [insert filename here]. You are advised to review the settings.'

cpnchecksumerror

**Radio Simulator**
Loaded the unhacked converted etx and launch the simulator from Companion and no error message - pass
Amended system settings and noticed the following debug info
54.19s: eeprom write general
54.19s: YAML radio settings writer
54.19s: generalSettings written with checksum 24840
54.19s: writeGeneralSettings error=SD card error
Manually confirmed the settings have NOT been saved as altered

Model changes are successfully written to the temp sd card folder
1.98s: eeprom write model
1.98s: YAML model writer
4.61s: [ViewMain] Focus lost
Manually confirmed the settings have been saved as altered

Due to error above could not test reloading amended and saved simulator files into Companion.

**Comment by stephenaa** [comment]
So the error to address is the "SD card error", right? After that your halted the testing since the file was not correct?

**Comment by elecpower** [comment]
> So the error to address is the "SD card error", right? After that your halted the testing since the file was not correct?

Yes and yes. Must have the ability to perform two way testing of reading and writing configuration.

**Comment by stephenaa** [comment]
Does this still have value? I have not had any time for EdgeTX work for a long time, and fell out of the routine :-(

**Comment by gagarinlg** [comment]
> Does this still have value? I have not had any time for EdgeTX work for a long time, and fell out of the routine :-(

Yes, it has some value, but I also did not find the time to take care of this PR

**Comment by gagarinlg** [comment]
@stephenaa Are you able to do a rebase of this PR by yourself or do you need help?

**Comment by gagarinlg** [comment]
The yaml_... files just have to be regenerated, so best is probably to keep the main state there when doing the rebase and not trying to merge them.
For the translations you have to do  some additional changes, becasue the way the translations are handled has changed, due to LVGL.

File: radio/src/storage/sdcard_yaml.cpp [hunk 7]
```diff
@@ -47,23 +49,74 @@ const char * readYamlFile(const char* fullpath, const YamlParserCalls* calls, vo
...
-    char buffer[32];
-    while (f_read(&file, buffer, sizeof(buffer), &bytes_read) == FR_OK) {
+    uint16_t calculated_checksum = 0xFFFF;
+    uint16_t file_checksum = 0;
 
-      // reached EOF?
-      if (bytes_read == 0)
+    bool first_block = true;
+    char buffer[32];
+    while (f_read(&file, buffer, sizeof(buffer)-1, &bytes_read) == FR_OK) {
```
[the remark this answers is not in the document]
Comment (gagarinlg) [reply radio/src/storage/sdcard_yaml.cpp:61]: yes, because of line 62, the loop is left when there a no bytes to read. 
There is the total sum of bytes read missing and that sum has to be used to fix this.

**Comment by gagarinlg** [comment]
I just did a rebase.
Following problems are not yet fixed:
- some translations are now missing, because they were still in the old format
- datastruct sizes are wrong because they changes in main and the PR
I will fix those issues in the next days

**Comment by gagarinlg** [comment]
> **Radio Simulator** Loaded the unhacked converted etx and launch the simulator from Companion and no error message - pass Amended system settings and noticed the following debug info 54.19s: eeprom write general 54.19s: YAML radio settings writer 54.19s: generalSettings written with checksum 24840 54.19s: writeGeneralSettings error=SD card error Manually confirmed the settings have NOT been saved as altered
> 
> Model changes are successfully written to the temp sd card folder 1.98s: eeprom write model 1.98s: YAML model writer 4.61s: [ViewMain] Focus lost Manually confirmed the settings have been saved as altered
> 
> Due to error above could not test reloading amended and saved simulator files into Companion.

The error message was wrong, everything was OK, but the return value of f_rename was not checked.
I could not find any problems using simu, except the the config write trigger does not come quick enough after leaving the page.

@elecpower @pfeerick could you two please also test, if this works as intended? And please be awar that it can take multiple seconds before changes settings are stored on the sd card.

**Comment by elecpower** [comment]
Retest error in X7 sim after clicking RE to clear wranings also when existing the sim the checksum changed
9.71s: YAML radio settings writer
9.71s: generalSettings written with checksum 41873
9.71s: writeGeneralSettings error=SD error
[D] targets/simu/opentxsimulator.cpp::OpenTxSimulator::stop():152 - ( 4388582418 us)
363.82s: opentxClose
363.82s: luaClose 0x7f03d4002160
363.82s: eeprom write general
363.82s: YAML radio settings writer
363.82s: generalSettings written with checksum 51435
363.82s: writeGeneralSettings error=SD error

Second attempt the first checksum same but second on exit changed from 1st attempt
11.39s: YAML radio settings writer
11.39s: generalSettings written with checksum 41873
11.39s: writeGeneralSettings error=SD error
[D] targets/simu/opentxsimulator.cpp::OpenTxSimulator::stop():152 - ( 4578618786 us)
21.82s: opentxClose
21.82s: luaClose 0x7f03e800be70
21.82s: eeprom write general
21.82s: YAML radio settings writer
21.82s: generalSettings written with checksum 30728
21.82s: writeGeneralSettings error=SD error
[

**Comment by gagarinlg** [comment]
@elecpower I could reproduce and fix the problem. Can you please retest again and when you are happy I would do a rebase.
Thx for your help

**Comment by elecpower** [comment]
@gagarinlg retested successfully. Write errors gone and read in both directions.

**Comment by gagarinlg** [comment]
> @gagarinlg retested successfully. Write errors gone and read in both directions.

Thank you very much.
I will do a rebase at the weekend.
@pfeerick @raphaelcoeffic do you have any additional comments on the code? If not, it should be OK to merge this, after a rebase.

**Comment by pfeerick** [comment]
On a partial skim read of the changes since I last look, seems fine. I'll try to play with this tomorrow. From what I can remember there were just the simulator and companion sync issues that were in need of resolving, so if they are gone all from me unless I find another loophole ;)

**Comment by gagarinlg** [comment]
> On a partial skim read of the changes since I last looked, seems fine. I'll try to play with this tomorrow. From what I can remember there were just the simulator and companion sync issues that were in need of resolving, so if they are gone all from me unless I find another loophole ;)

I just found somethinig while doing a test run. There have been corner cases with specific yaml file length when converting from without CRC to with CRC on the radio. I luckily hit one of those.
A fix is already pushed

**Comment by pfeerick** [comment]
With Zorro as the unlucky duck, so far the initial migration from non-CRC to CRC was seamless. `manuallyEdited` let me tweak a setting without issue. Failing to do so triggered error recovery (and since I knew the change I made that broke it, revering the change, and moving the file back worked also). Removing the checksum block as another workaround was also good. My only gripe so far would be translations strings need a shorter alternative - current string `TR_BAD_RADIO_DATA_UNRECOVERABLE` => "Unable to recover radio data" misses the last word on 128x64 screen, and I expect `TR_RADIO_DATA_RECOVERED` will be worse. 

Since it seems there is a recovery mode based on `radio_new.yml` (under what circumstances is `radio_new.yml` expected to exist?), I then copied `radio.yml` to `radio_new.yml`, changed `radio.yml` so it would break, and it successfully recovered (and renamed the files in place leaving just the now "suspect" `radio_error.yml`. Editing the radio settings with Companion was also successful. 

I can sort the translation strings out in the morning when I play with widescreen and colorlcd, but it's looking pretty merge-worthy so far ;)

**Comment by gagarinlg** [comment]
> Since it seems there is a recovery mode based on `radio_new.yml` (under what circumstances is `radio_new.yml` expected to exist?), I then copied `radio.yml` to `radio_new.yml`, changed `radio.yml` so it would break, and it successfully recovered (and renamed the files in place leaving just the now "suspect" `radio_error.yml`. Editing the radio settings with Companion was also successful.
> 
The config is first written to radio_new.yml. When this is successful, radio.yml is deleted and radio_new.yml renamed.
When the radio is interrupted while writing radio_new.yml, radio.yml is still OK and will be used. When the radio is interrupted after deleting radio.yml, before renaming radio_new.yml, radio_new.yml is used.
so there should be no time window to create a broken configuration, because either radio.yml or radio_new.yml should be OK.
When the currently existing config (radio.yml, or radio_new.yml) can not be read, it is renamned to radio_error.yml, so tha a post mortem analysis of the broken file is possible

Do what ever you want with the strings.

**Review by pfeerick — APPROVED** [review]
I've now played with this on TX16S, TX12, Zorro, NV14 and X9D+2019.

- editing `radio.yml` without altering `manuallyEdited` consistently triggered `RADIO_DATA_UNRECOVERABLE `
- editing `radio.yml` and altering `manuallyEdited` would start normally
- editing with companion would start normally
- renaming `radio.yml` to `radio_new.yaml` would trigger `RADIO_DATA_RECOVERED `
- repeated restarts on X9D+2019 (which was one of the radios that has had settings reset issue in the past - although infrequent for me) did not trigger the warning (nor were settings lost) AFAICT

I have updated the translations strings to have three strings for the two prompts, to suit 128, 212 and long screen size. These shouldn't be translated until we're happy with the wording, just in case the wording I use was felt to be misleading or otherwise wrong. It was certainly tricky trying not to be too cryptic within the screen widths. 

Assuming no build failues due to my changes, this should be in the next nightly. I do have some requests, suggestions for future workflow of this - as it seems a bit annoying that it will say unrecoverabled, and then bad radio data, and then storage prep - it would be a nicer end user experience if it skipped at least the middle prompt. But I think this is for another PR as isn't the focus of this PR.
